File: my_flutter_app/lib/services/notification_routing.dart
import '../screens/review_composer_screen.dart';
import '../screens/tracking_screen.dart';

/// Where to navigate for a given push notification's `data` payload.
class NotificationRoute {
  const NotificationRoute(this.name, {this.arguments});

  final String name;
  final Object? arguments;
}

/// Focus for the customer Orders screen (inner tab + record).
///
/// Mirrors `customerOrdersInnerTab` / `customerTargetForNotification`
/// in admin_web so a bell tap opens the same shipment the web console does.
class OrdersScreenArguments {
  const OrdersScreenArguments({
    this.innerTab,
    this.focusId,
    this.focusCollection,
  });

  /// `barrels` | `freight` | `transport` | `cars`
  final String? innerTab;
  final String? focusId;
  final String? focusCollection;
}

const _typeCollections = <String, String>{
  'barrel_shipment_status': 'barrelShipments',
  'freight_shipment_status': 'freightShipments',
  'freight_balance_due': 'freightShipments',
  'freight_refund_issued': 'freightShipments',
  'freight_quote_received': 'freightQuoteRequests',
  'transport_request_status': 'transportRequests',
  'parking_reservation_status': 'parkedCars',
  'car_purchase_status': 'carPurchases',
  'car_viewing_status': 'carPurchases',
};

String collectionForNotificationType(String type) =>
    _typeCollections[type] ?? '';

String recordIdFromNotification(Map<String, dynamic> data) {
  for (final key in [
    'relatedId',
    'shipmentId',
    'requestId',
    'reservationId',
    'purchaseId',
    'recordId',
  ]) {
    final value = data[key]?.toString().trim() ?? '';
    if (value.isNotEmpty) return value;
  }
  return '';
}

String collectionFromNotification(Map<String, dynamic> data) {
  final stamped = data['relatedCollection']?.toString().trim() ?? '';
  if (stamped.isNotEmpty) return stamped;
  return collectionForNotificationType(data['type']?.toString() ?? '');
}

String customerOrdersInnerTab(String collection) {
  switch (collection) {
    case 'barrelShipments':
    case 'barrelOrders':
      return 'barrels';
    case 'freightShipments':
    case 'freightQuoteRequests':
      return 'freight';
    case 'transportRequests':
      return 'transport';
    default:
      return 'cars';
  }
}

NotificationRoute _ordersRoute(Map<String, dynamic> data) {
  final collection = collectionFromNotification(data);
  final id = recordIdFromNotification(data);
  return NotificationRoute(
    '/orders',
    arguments: OrdersScreenArguments(
      innerTab: customerOrdersInnerTab(collection),
      focusId: id.isEmpty ? null : id,
      focusCollection: collection.isEmpty ? null : collection,
    ),
  );
}

/// Best-effort deep link: routes to the most relevant screen for this
/// notification's category. Several detail screens (parked car, barrel
/// shipment, transport request) require a full loaded model as a route
/// argument rather than just an id, so this opens the closest list/hub
/// screen instead of fetching and constructing the full object here.
/// Returns null when the type is unrecognized or required data is missing.
NotificationRoute? routeForNotificationData(Map<String, dynamic> data) {
  final type = data['type']?.toString() ?? '';
  switch (type) {
    case 'car_purchase_status':
      return const NotificationRoute('/my-purchases');
    case 'car_viewing_status':
      // Viewings have their own destination: a buyer opening a notification
      // about an appointment should not land in a list of purchases.
      return const NotificationRoute('/my-viewings');
    case 'barrel_shipment_status':
    case 'freight_shipment_status':
    case 'freight_balance_due':
    case 'freight_refund_issued':
    case 'parking_reservation_status':
    case 'transport_request_status':
      return _ordersRoute(data);
    // A price landing is the one notification a customer is actually waiting
    // on, and it used to open nothing: there was no case here and no named
    // route to send it to, so the tap fell through and the customer had no
    // way back to the price they had just been told about.
    case 'freight_quote_received':
      final requestId = data['requestId']?.toString() ?? '';
      if (requestId.isEmpty) return null;
      return NotificationRoute(
        '/freight-quote',
        arguments: FreightQuoteScreenArguments(
          requestId: requestId,
          trackingCode: data['trackingCode']?.toString() ?? '',
        ),
      );
    case 'review_request':
      final relatedCollection = data['relatedCollection']?.toString();
      final relatedId = data['relatedId']?.toString();
      final businessId = data['businessId']?.toString();
      if (relatedCollection == null ||
          relatedCollection.isEmpty ||
          relatedId == null ||
          relatedId.isEmpty ||
          businessId == null ||
          businessId.isEmpty) {
        return null;
      }
      return NotificationRoute(
        '/leave-review',
        arguments: ReviewComposerArguments(
          relatedCollection: relatedCollection,
          relatedId: relatedId,
          businessId: businessId,
        ),
      );
    // No wallet_refund_status case: the wallet is retired
    // (docs/PLAN-2026-08-backlog.md #3). The backend never sends that type and
    // /wallet is not a registered route, so it could only ever have been a tap
    // that did nothing.
    case 'support_case_update':
      // Same payload and destination as support_message - the backend sends
      // this on every case update (index.js:25176) and it had no route, so
      // those taps silently went nowhere.
      final supportCaseId = data['caseId']?.toString();
      if (supportCaseId == null || supportCaseId.isEmpty) return null;
      return NotificationRoute('/support-thread', arguments: supportCaseId);
    case 'shipment_tracking_update':
      return NotificationRoute(
        '/tracking',
        arguments: TrackingScreenArguments(
          shipmentId: data['shipmentId']?.toString(),
        ),
      );
    case 'business_application_status':
      return const NotificationRoute('/business-register');
    case 'support_message':
    case 'support_escalated':
      final caseId = data['caseId']?.toString();
      if (caseId == null || caseId.isEmpty) return null;
      return NotificationRoute('/support-thread', arguments: caseId);
    default:
      return null;
  }
}

/// FCM data payloads are always string-keyed and string-valued, so encoding
/// as a query string is enough to round-trip them through
/// flutter_local_notifications' payload field (used when a
/// foreground-shown local notification is
/// tapped, since that path doesn't hand back the original RemoteMessage).
String encodeNotificationPayload(Map<String, dynamic> data) {
  return data.entries
      .map(
        (entry) =>
            '${Uri.encodeComponent(entry.key)}='
            '${Uri.encodeComponent(entry.value.toString())}',
      )
      .join('&');
}

Map<String, String> decodeNotificationPayload(String payload) {
  final data = <String, String>{};
  if (payload.isEmpty) return data;
  for (final pair in payload.split('&')) {
    final separatorIndex = pair.indexOf('=');
    if (separatorIndex <= 0) continue;
    final key = Uri.decodeComponent(pair.substring(0, separatorIndex));
    final value = Uri.decodeComponent(pair.substring(separatorIndex + 1));
    data[key] = value;
  }
  return data;
}

/// What a notification about a price carries to the screen that shows it.
class FreightQuoteScreenArguments {
  const FreightQuoteScreenArguments({
    required this.requestId,
    this.trackingCode = '',
  });

  final String requestId;
  final String trackingCode;
}

File: admin_web/src/lib/notification-routing.ts
/**
 * Website notification click routing.
 *
 * Mirrors the mobile contract in
 * `my_flutter_app/lib/services/notification_routing.dart`, then lands on the
 * matching console tab and record. Live payloads often carry only a typed id
 * (`shipmentId`, `requestId`, …) and no `relatedCollection`; this module
 * infers the collection from `type` so an already-sent notification still
 * opens the right inner tab.
 */

import type { BusinessTab } from "@/lib/business-sidebar";

export type NotificationData = Record<string, string>;

export type NotificationFocus = {
  collection: string;
  id: string;
};

export type CustomerNotificationTab =
  | "home"
  | "services"
  | "parkingPools"
  | "cars"
  | "viewings"
  | "orders"
  | "support"
  | "profile";

export type CustomerNotificationTarget = {
  tab: CustomerNotificationTab;
  focus?: NotificationFocus;
  caseId?: string;
  openReview?: boolean;
};

export type BusinessPanelView =
  | "opportunities"
  | "jobs"
  | "requests"
  | "shipments";

export type BusinessNotificationTarget = {
  tab: BusinessTab;
  focusId?: string;
  panelView?: BusinessPanelView;
  caseId?: string;
};

export type AdminNotificationTarget = {
  tab: string;
  caseId?: string;
};

const ORDER_TYPE_COLLECTIONS: Record<string, string> = {
  barrelShipment: "barrelShipments",
  barrelOrder: "barrelOrders",
  barrelDestinationChange: "barrelShipments",
  freightShipment: "freightShipments",
  freightSettlement: "freightShipments",
  transportJob: "transportRequests",
  transport_job: "transportRequests",
  parking: "parkedCars",
  parking_deposit: "parkedCars",
  carDeposit: "carPurchases",
  carPurchase: "carPurchases",
  holdExtension: "carPurchases",
  reservation_deposit: "carPurchases",
  full_purchase: "carPurchases",
};

const TYPE_COLLECTIONS: Record<string, string> = {
  barrel_shipment_status: "barrelShipments",
  freight_shipment_status: "freightShipments",
  freight_balance_due: "freightShipments",
  freight_refund_issued: "freightShipments",
  freight_quote_received: "freightQuoteRequests",
  freight_quote_request: "freightQuoteRequests",
  freight_quote_won: "freightQuoteRequests",
  freight_quote_lost: "freightQuoteRequests",
  transport_request_status: "transportRequests",
  transport_opportunity: "transportRequests",
  transport_quote_won: "transportRequests",
  transport_quote_lost: "transportRequests",
  transport_job_paid: "transportRequests",
  parking_reservation_status: "parkedCars",
  car_purchase_status: "carPurchases",
  car_viewing_status: "carPurchases",
  review_request: "",
  shipment_tracking_update: "",
  payment_hold_capture_notice: "",
  secured_order_cancelled_by_business: "",
  barrel_pool_deposit: "barrelPools",
  barrel_pool_join: "barrelPools",
  barrel_pool_balance_due: "barrelPools",
  deposit_refund_due: "barrelPools",
};

function firstText(...values: Array<string | undefined>): string {
  for (const value of values) {
    const text = String(value ?? "").trim();
    if (text) return text;
  }
  return "";
}

export function collectionForOrderType(orderType: string): string {
  return ORDER_TYPE_COLLECTIONS[orderType] ?? "";
}

export function collectionForNotificationType(type: string): string {
  return TYPE_COLLECTIONS[type] ?? "";
}

export function recordIdFromNotification(data: NotificationData): string {
  return firstText(
    data.relatedId,
    data.shipmentId,
    data.requestId,
    data.reservationId,
    data.purchaseId,
    data.recordId,
  );
}

export function collectionFromNotification(data: NotificationData): string {
  const stamped = firstText(data.relatedCollection);
  if (stamped) return stamped;
  const fromType = collectionForNotificationType(data.type ?? "");
  if (fromType) return fromType;
  return collectionForOrderType(firstText(data.orderType, data.service));
}

export function focusFromNotification(
  data: NotificationData,
): NotificationFocus | undefined {
  const id = recordIdFromNotification(data);
  if (!id) return undefined;
  return {collection: collectionFromNotification(data), id};
}

export function caseIdFromNotification(data: NotificationData): string {
  return firstText(data.caseId);
}

/**
 * Where a customer notification click should land.
 *
 * Live shipment/status payloads often have `shipmentId` and no
 * `relatedCollection`. Collection is inferred from `type` so Orders opens
 * Barrels/Freight instead of the leftover Cars tab.
 */
export function customerTargetForNotification(
  data: NotificationData,
): CustomerNotificationTarget {
  const type = data.type ?? "";
  const focus = focusFromNotification(data);
  const caseId = caseIdFromNotification(data);

  switch (type) {
    case "support_message":
    case "support_escalated":
    case "support_case_update":
      return caseId ? {tab: "support", caseId} : {tab: "support"};
    case "car_viewing_status":
      return {tab: "viewings", focus};
    case "car_purchase_status":
    case "barrel_shipment_status":
    case "freight_shipment_status":
    case "freight_balance_due":
    case "freight_refund_issued":
    case "freight_quote_received":
    case "parking_reservation_status":
    case "transport_request_status":
    case "shipment_tracking_update":
    case "deposit_refund_due":
    case "payment_hold_capture_notice":
    case "secured_order_cancelled_by_business":
      return {tab: "orders", focus};
    case "review_request":
      return {tab: "orders", focus, openReview: true};
    case "barrel_pool_deposit":
    case "barrel_pool_join":
    case "barrel_pool_balance_due":
      return {tab: "parkingPools", focus};
    case "business_application_status":
      return {tab: "profile"};
    default:
      return {tab: "home"};
  }
}

function businessPaidTab(service: string): BusinessTab {
  switch (service) {
    case "barrels":
      return "barrels";
    case "freight":
      return "freight";
    case "transport":
      return "transport";
    case "parking":
      return "parking";
    case "purchases":
      return "purchases";
    default:
      return "today";
  }
}

/**
 * Where a business notification click should land, including the inner
 * panel view (Accepted jobs vs Opportunities, price-requests vs shipments).
 */
export function businessTargetForNotification(
  data: NotificationData,
): BusinessNotificationTarget {
  const type = data.type ?? "";
  const service = data.service ?? "";
  const focusId = recordIdFromNotification(data);
  const caseId = caseIdFromNotification(data);

  if (type === "business_order_paid") {
    const tab = businessPaidTab(service);
    return {
      tab,
      focusId,
      panelView:
        service === "transport"
          ? "jobs"
          : service === "freight"
            ? "shipments"
            : undefined,
    };
  }

  switch (type) {
    case "support_message":
    case "support_escalated":
    case "support_case_update":
      return {tab: "cases", caseId, focusId: caseId};
    case "business_verification_review":
    case "business_verification_document":
    case "business_application_status":
      return {tab: "profile"};
    case "barrel_shipment_status":
      return {tab: "barrels", focusId};
    case "freight_shipment_status":
    case "freight_balance_due":
    case "freight_refund_issued":
      return {tab: "freight", focusId, panelView: "shipments"};
    case "freight_quote_request":
    case "freight_quote_won":
    case "freight_quote_lost":
      return {tab: "freight", focusId, panelView: "requests"};
    case "parking_reservation_status":
      return {tab: "parking", focusId};
    case "car_viewing_status":
      return {tab: "viewings", focusId};
    case "car_purchase_status":
      return {tab: "purchases", focusId};
    case "transport_opportunity":
    case "transport_quote_lost":
      return {tab: "transport", focusId, panelView: "opportunities"};
    case "transport_request_status":
    case "transport_quote_won":
    case "transport_job_paid":
      return {tab: "transport", focusId, panelView: "jobs"};
    default:
      return {tab: "today"};
  }
}

export function adminTargetForNotification(
  data: NotificationData,
): AdminNotificationTarget {
  const type = data.type ?? "";
  const caseId = caseIdFromNotification(data);
  switch (type) {
    case "support_message":
    case "support_escalated":
    case "support_case_update":
      return {tab: "support", caseId};
    case "business_application":
    case "business_application_status":
    case "business_verification_document":
    case "business_verification_review":
      return {tab: "businesses"};
    default:
      return {tab: "today"};
  }
}

/** Orders inner-tab for a focused customer record. */
export function customerOrdersInnerTab(
  collection: string,
): "barrels" | "freight" | "transport" | "cars" {
  switch (collection) {
    case "barrelShipments":
      return "barrels";
    case "freightShipments":
    case "freightQuoteRequests":
      return "freight";
    case "transportRequests":
      return "transport";
    default:
      return "cars";
  }
}
